### PR TITLE
serve multiple thrift ifaces

### PR DESCRIPTION
**Pull request by thirstycrow**
Problem

Finagle uses an early version of libthrift, which does not support
service multiplexing.

Solution

Since 0.9.1, thrift added several classes to handle service multiplexing.
On the client side, TMultiplexedProtocol is used to decorate the
original protocol, by prefixing a user specified service name as part of
the name of request message. Finagle uses the protocols from libthrift.
All we have to do is copying TMultiplexedProtocol and related classes
from a newer version of libthrfit.

On the server side, TMultiplexedProcessor is used to inspect the
beginning of the incoming message names, and dispatch the message
to an appropriate underlying processor. Finagle does not use
processors, so we have to re-implement this.

Result

Multiple thrift services can be served through a single port, by passing
a map from service names to service implementations.

``` scala
val serviceMap = Map(
  "echo" -> new EchoService(),
  "extendedEcho" -> new ExtendedEchoService()
)

val server = Thrift.server.serveIfaces(address, serviceMap)
```

Multiple Ifaces/ServiceIfaces can be created sharing the same underlying
thrift client.

``` scala
val client = Thrift.client.multiplex(address, "client") { client =>
  new {
    val echo = client.newIface[Echo.FutureIface]("echo")
    val extendedEcho = client.newServiceIface[ExtendedEcho.ServiceIface]("extendedEcho")
  }
}

client.echo.echo("hello")
client.extendedEcho.getStatus(ExtendedEcho.GetStatus.Args())
```

### CODE REVIEW

**Comment by kevinoliver** [comment]
@thirstycrow this sounds nice and I know there have been teams here at Twitter who would like something like this (they've worked around it in a different way, but I can't recall the details).

I'll try to take a look at your change this week.

File: finagle-thrift/src/main/scala/com/twitter/finagle/rich.scala [hunk 1]
```diff
@@ -380,7 +396,18 @@ trait ThriftRichClient { self: Client[ThriftClientRequest, Array[Byte]] =>
...
+
+  /**
+   * $clientUse
+   */
+  def newIface[Iface](
+    name: Name,
+    label: String,
+    cls: Class[_],
+    protocolFactory: TProtocolFactory,
+    service: Service[ThriftClientRequest, Array[Byte]]): Iface = {
```
Comment (kevinoliver) [inline finagle-thrift/src/main/scala/com/twitter/finagle/rich.scala]: style nitpick: move the `): Iface = {` down to the next line, indented to the same as the `def`.

File: finagle-thrift/src/main/scala/com/twitter/finagle/rich.scala [hunk 2]
```diff
@@ -447,6 +473,43 @@ trait ThriftRichClient { self: Client[ThriftClientRequest, Array[Byte]] =>
   def newMethodIface[ServiceIface, FutureIface](serviceIface: ServiceIface)(
     implicit builder: MethodIfaceBuilder[ServiceIface, FutureIface]
   ): FutureIface = builder.newMethodIface(serviceIface)
+
+  def multiplex[T](dest: Name, label: String)(build: MultiplexedThriftClient => T): T = {
```
Comment (kevinoliver) [inline finagle-thrift/src/main/scala/com/twitter/finagle/rich.scala]: lets do some scaladocs for these new `multiplex` methods.

File: finagle-thrift/src/main/scala/com/twitter/finagle/rich.scala [hunk 3]
```diff
@@ -447,6 +473,43 @@ trait ThriftRichClient { self: Client[ThriftClientRequest, Array[Byte]] =>
...
+    build(new MultiplexedThriftClient(dest, label))
+  }
+
+  def multiplex[T](dest: String, label: String)(build: MultiplexedThriftClient => T): T = {
+    multiplex(Resolver.eval(dest), label)(build)
+  }
+
+  def multiplex[T](dest: String)(build: MultiplexedThriftClient => T): T = {
```
Comment (kevinoliver) [inline finagle-thrift/src/main/scala/com/twitter/finagle/rich.scala]: mind omitting this and forcing these clients to specify a label? its for their own good ;)

File: finagle-thrift/src/main/scala/com/twitter/finagle/rich.scala [hunk 4]
```diff
@@ -447,6 +473,43 @@ trait ThriftRichClient { self: Client[ThriftClientRequest, Array[Byte]] =>
...
+    val (n, l) = Resolver.evalLabeled(dest)
+    multiplex(n, l)(build)
+  }
+
+  class MultiplexedThriftClient(dest: Name, label: String) {
+
+    val service = newService(dest, label)
```
Comment (kevinoliver) [inline finagle-thrift/src/main/scala/com/twitter/finagle/rich.scala]: can you reduce the visibility?

File: finagle-thrift/src/main/scala/com/twitter/finagle/thrift/MultiplexedService.scala [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+import com.twitter.util.Future
+import java.util.Arrays
+import org.apache.thrift.TApplicationException
+import org.apache.thrift.protocol._
+import org.apache.thrift.transport.TMemoryInputTransport
+import scala.collection.JavaConverters._
+import scala.util.Try
+
+class MultiplexedFinagleService(
+  services: Map[String, Service[Array[Byte], Array[Byte]]],
```
Comment (kevinoliver) [inline finagle-thrift/src/main/scala/com/twitter/finagle/thrift/MultiplexedService.scala]: style nit, format for multi line class declarations:

```
class MyName(
    arg1: Type1,
    ...)
  extends ... {
```

File: finagle-thrift/src/main/scala/com/twitter/finagle/thrift/MultiplexedService.scala [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+import com.twitter.scrooge.TReusableMemoryTransport
+import com.twitter.util.Future
+import java.util.Arrays
+import org.apache.thrift.TApplicationException
+import org.apache.thrift.protocol._
+import org.apache.thrift.transport.TMemoryInputTransport
+import scala.collection.JavaConverters._
+import scala.util.Try
+
+class MultiplexedFinagleService(
```
Comment (kevinoliver) [inline finagle-thrift/src/main/scala/com/twitter/finagle/thrift/MultiplexedService.scala]: can you reduce visibility?

File: finagle-thrift/src/main/scala/com/twitter/finagle/thrift/MultiplexedService.scala [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+    extends Service[Array[Byte], Array[Byte]] {
+
+  private val serviceMap = services.mapValues(getFunctionMap)
+
+  def getFunctionMap(
+    service: Service[Array[Byte], Array[Byte]]
+  ): collection.Map[String, (TProtocol, Int) => Future[Array[Byte]]] = {
+    Try(scroogeFinagleServiceFunctionMap(service))
+      .orElse(Try(thriftFinagleServiceFunctionMap(service)))
+      .getOrElse(throw new IllegalArgumentException("%s cannot be multiplexed".format(service.getClass.getName)))
```
Comment (kevinoliver) [inline finagle-thrift/src/main/scala/com/twitter/finagle/thrift/MultiplexedService.scala]: might be helpful to chain the Throw's exception in the failure case

File: finagle-thrift/src/main/scala/com/twitter/finagle/thrift/MultiplexedService.scala [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+  }
+
+  def scroogeFinagleServiceFunctionMap[T](target: AnyRef) = {
```
Comment (kevinoliver) [inline finagle-thrift/src/main/scala/com/twitter/finagle/thrift/MultiplexedService.scala]: why do we need to support both this and `thriftFinagleServiceFunctionMap`? (mind leaving a comment in the code as a response?)

File: finagle-thrift/src/main/scala/com/twitter/finagle/thrift/MultiplexedService.scala [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+            val errorMsg = if (functionMap.isEmpty) {
+              s"""Invalid service name: ${serviceName}, available serivce names: [${serviceMap.keys.mkString(", ")}]"""
+            } else {
+              s"""Invalid method name: ${functionName}, available method names: [${functionMap.get.keys.mkString(", ")}]"""
+            }
+            exception(msg.name, msg.seqid, TApplicationException.UNKNOWN_METHOD, errorMsg)
+        }
+      }
+    } catch {
+      case e: Throwable => Future.exception(e)
```
Comment (kevinoliver) [inline finagle-thrift/src/main/scala/com/twitter/finagle/thrift/MultiplexedService.scala]: should probably only catch NonFatal's:

`case NonFatal(e) => Future.exception(e)`

File: finagle-thrift/src/main/scala/com/twitter/finagle/thrift/MultiplexedService.scala [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+        oprot.writeMessageBegin(new TMessage(name, TMessageType.EXCEPTION, seqid))
+        x.write(oprot)
+        oprot.writeMessageEnd()
+        oprot.getTransport().flush()
+        Future.value(Arrays.copyOfRange(memoryBuffer.getArray(), 0, memoryBuffer.length()))
+      } finally {
+        resetBuffer(memoryBuffer)
+      }
+    } catch {
+      case e: Exception => Future.exception(e)
```
Comment (kevinoliver) [inline finagle-thrift/src/main/scala/com/twitter/finagle/thrift/MultiplexedService.scala]: ditto re NonFatal

File: finagle-thrift/src/main/scala/com/twitter/finagle/thrift/Protocols.scala [hunk 11]
```diff
@@ -78,6 +78,17 @@ object Protocols {
     binaryFactory(statsReceiver = statsReceiver)
   }
 
+  /**
+   * Multiplex the protocolFactory with the serviceName.
+   */
+  def multiplex(serviceName: String, protocolFactory: TProtocolFactory) = {
```
Comment (kevinoliver) [inline finagle-thrift/src/main/scala/com/twitter/finagle/thrift/Protocols.scala]: for public methods, annotate with the return type

File: finagle-thrift/src/main/scala/com/twitter/finagle/thrift/Protocols.scala [hunk 12]
```diff
@@ -78,6 +78,17 @@ object Protocols {
     binaryFactory(statsReceiver = statsReceiver)
   }
 
+  /**
+   * Multiplex the protocolFactory with the serviceName.
```
Comment (kevinoliver) [inline finagle-thrift/src/main/scala/com/twitter/finagle/thrift/Protocols.scala]: explain why would you use this? i'm guessing its when you have a client that is talking to a multiplexed backend service?

**Comment by kevinoliver** [comment]
also, apologies for the delay in reviewing. it got lost in my inbox.

**Comment by thirstycrow** [comment]
@kevinoliver I'm not quite familiar with the SHA thing. Could you add it before merging this pull request? The libthrift code is copied exactly from [the 0.9.1 tag](https://github.com/apache/thrift/tree/0.9.1)

I added a param to specify a default service，so the server can be upgraded to use mutiplexing, and still be compatible with the old clients.
```scala
val server = Thrift.server.serveIfaces(
  address, serviceMap, defaultService = Some("extendedEcho"))
```

**Comment by taylorleese** [comment]
What (if anything) needs to be done for this to also work with ThriftMux?

**Comment by kevinoliver** [comment]
> I'm not quite familiar with the SHA thing. Could you add it before merging this pull request? The libthrift code is copied exactly from the 0.9.1 tag

a link to that tag works perfectly for me. the idea is just to help future devs track down exactly where the code originated from.

**Comment by kevinoliver** [comment]
the default service addition is nice. thanks.

**Comment by thirstycrow** [comment]
@taylorleese I think there's nothing special needs to be done to work with ThriftMux, which does not depend on the details of how a thrift message is encoded.

**Comment by kevinoliver** [comment]
this lgtm. be good to have another set of eyes on it.

**Comment by mosesn** [comment]
@thirstycrow sorry for the delay, this looks good to me!  I'll see if I can merge it in.

**Comment by mosesn** [comment]
@thirstycrow it's gonna be a bit longer, we need to get legal to sign off on copying the code from Apache Thrift in.

**Comment by mosesn** [comment]
@thirstycrow sorry for the delay, our lawyers finally got back to us.  this was merged in! https://github.com/twitter/finagle/commit/b9cb4883ce35e4f5a88c37d0fc4eeb507ddda0a0

**Pull request closed without merging**